modules/ts/tse/functions.cpp

```diff
@@ -0,0 +1,519 @@
+	if (!IsPlayerValid(amx, pid)) return 0;
+	if (!Player(pid)->IsAlive()) return 0;
+	byte weapon = params[2];
+	byte attachments = params[3];
+	byte isactive = params[4];
+	if (weapon < 1 || weapon > MAXWEAPS) return 0;
+	if (attachments < 0 || attachments > 15) return 0;
+	if (!WeaponsList[weapon].offsets.clip) return 0;
+	if (Player(pid)->CurrentWeapon == weapon)
+	{
+		Player(pid)->Weapons[weapon].attachments = attachments;
+		Player(pid)->UpdateHUD();
+		if (isactive) {
+			MESSAGE_BEGIN(MSG_ONE, GetMsgIDByName("ActItems"), 0, Player(pid)->PlayerEdict);
+				WRITE_BYTE(1);
+				WRITE_BYTE(attachments);
+			MESSAGE_END();
+		}
+	}
+	if (isactive) {
+		Player(pid)->SetWeapPDataInt(WeaponsList[weapon].offsets.clip + 3, attachments);
+		Player(pid)->SetWeapPDataInt(WeaponsList[weapon].offsets.clip + 4, attachments);
+	}
+	else
+		Player(pid)->SetWeapPDataInt(WeaponsList[weapon].offsets.clip + 3, attachments);
+	return 1;
+}
+
+static cell AMX_NATIVE_CALL tse_setweapclip(AMX *amx, cell *params)
+{
+	byte pid = params[1];
+	if (!IsPlayerValid(amx, pid)) return 0;
+	if (!Player(pid)->IsAlive()) return 0;
+	byte weapon = params[2];
+	if (!WeaponsList[weapon].offsets.clip) return 0;
+	byte clip = params[3];
+	Player(pid)->SetWeapPDataInt(WeaponsList[weapon].offsets.clip, clip);
+	return 1;
+}
+
+static cell AMX_NATIVE_CALL tse_setweapammo(AMX *amx, cell *params)
+{
+	byte pid = params[1];
+	if (!IsPlayerValid(amx, pid)) return 0;
+	if (!Player(pid)->IsAlive()) return 0;
+	byte weapon = params[2];
+	if (!WeaponsList[weapon].offsets.clip) return 0;
+	byte ammo = params[3];
+	Player(pid)->SetWeapPDataInt(WeaponsList[weapon].offsets.ammo, ammo);
+	return 1;
+}
+
+static cell AMX_NATIVE_CALL tse_setweapready(AMX *amx, cell *params)
+{
+	byte pid = params[1];
+	if (!IsPlayerValid(amx, pid)) return 0;
+	if (!Player(pid)->IsAlive()) return 0;
+	byte weapon = params[2];
+	if (!WeaponsList[weapon].offsets.clip) return 0;
+	int isready = params[3];
+	Player(pid)->SetWeapPDataInt(WeaponsList[weapon].offsets.clip + 8, (int)!(bool)(isready != 0));
+	return 1;
+}
+
+static cell AMX_NATIVE_CALL tse_getweapinfo(AMX *amx, cell *params)
+{
+	byte pid = params[1];
+	if (!IsPlayerValid(amx, pid)) return 0;
+	if (!Player(pid)->IsAlive()) return 0;
+	byte weapon = params[2];
+	if (weapon < 1 || weapon > MAXWEAPS) return 0;
+	if (!WeaponsList[weapon].offsets.clip) return 0;
+	cell *clip = MF_GetAmxAddr(amx, params[3]);
+	cell *ammo = MF_GetAmxAddr(amx, params[4]);
+	cell *attachs = MF_GetAmxAddr(amx, params[5]);
+	cell *active_attachs = MF_GetAmxAddr(amx, params[6]);
+	*clip = Player(pid)->GetWeapPDataInt(WeaponsList[weapon].offsets.clip);
+	*ammo = Player(pid)->GetWeapPDataInt(WeaponsList[weapon].offsets.ammo);
+	*attachs = Player(pid)->GetWeapPDataInt(WeaponsList[weapon].offsets.clip + 3);
+	*active_attachs = Player(pid)->GetWeapPDataInt(WeaponsList[weapon].offsets.clip + 4);
+	return 1;
+}
+
+static cell AMX_NATIVE_CALL tse_isuserhasweap(AMX *amx, cell *params)
```

Comment:
This function can be refactored to something like this.

```
static cell AMX_NATIVE_CALL tse_isuserhasweap(AMX *amx, cell *params)
{
    byte pid = params[1];
    
    if (!IsPlayerValid(amx, pid)) return -1;
    if (!Player(pid)->IsAlive()) return -1;
    
    byte weapon = params[2];
    
    if (!WeaponsList[weapon].offsets.clip) return -1;
    
    byte offset = !!WeaponsList[weapon].offsets.fmbase;
    cell result = Player(pid)->GetWeapPDataInt(WeaponsList[weapon].offsets.clip - offset) > 0;
    
    return result;
}
```

Reply:
@In-line I prefer explicit:
`byte offset = !!WeaponsList[weapon].offsets.fmbase;`
->
`byte offset = WeaponsList[weapon].offsets.fmbase != 0 ? 1 : 0;`

Reply:
I applied changes in this aspect, thanks.